### ISSUE TITLE
Can't use x:Bind with Converter in Window

### ISSUE THREAD

**Issue by alexdi220** [microsoft/microsoft-ui-xaml#4966]
**Describe the bug**
Hi, I've tried to use the x:Bind with IValueConverter. But the project doesn't compile - `CS1503	Argument 1: cannot convert from 'TestApp.MainWindow' to 'Microsoft.UI.Xaml.FrameworkElement'`. Please take a look at the following code. What I do wrong?

**Steps to reproduce the bug**
```xaml
<Window
    x:Class="TestApp.MainWindow"
    xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
    xmlns:local="using:TestApp"
    xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
    xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
    mc:Ignorable="d">
    <Grid RowDefinitions="Auto, *" x:Name="root">
        <Grid.Resources>
            <local:MyConverter x:Key="MyConverter"/>
        </Grid.Resources>
        <StackPanel Spacing="12"
                    Grid.Row="1">
            <TextBlock Text="{x:Bind MyProperty, Converter={StaticResource MyConverter}}"/>
        </StackPanel>
    </Grid>
</Window>
```
```csharp
    public class MyConverter : IValueConverter {
        public object Convert(object value, Type targetType, object parameter, string language) => value;
        public object ConvertBack(object value, Type targetType, object parameter, string language) => value;
    }

    public sealed partial class MainWindow : Window {
        public MainWindow() {
            this.InitializeComponent();            
        }
        public string MyProperty { get; } = "Test";
```

**Expected behavior**

**Screenshots**
![image](https://user-images.githubusercontent.com/47348109/117275767-0c877300-ae67-11eb-8130-69d0aa5a93b2.png)

**Version Info**

NuGet package version: 

[WinUI 3 - Project Reunion 0.5: 0.5.6]

Windows app type:
| UWP              | Win32            |
| :--------------- | :--------------- |
|  | Yes|

| Windows 10 version                  | Saw the problem? |
| :--------------------------------- | :-------------------- |
| Insider Build (xxxxx)              |    |
| October 2020 Update (19042)        | Yes   |
| May 2020 Update (19041)            |    |
| November 2019 Update (18363)       |    |
| May 2019 Update (18362)            |    |
| October 2018 Update (17763)        |    |
| April 2018 Update (17134)          |    |
| Fall Creators Update (16299)       |    |
| Creators Update (15063)            |    |

| Device form factor | Saw the problem? |
| :----------------- | :--------------- |
| Desktop            |  Yes |
| Xbox               |  |
| Surface Hub        |  |
| IoT                |  |

**Additional context**

**Comment by ranjeshj** [comment]
@RealTommyKlein FYI

**Comment by RealTommyKlein** [comment]
Thanks for the report.  The root of your `.xaml` needs to be a `FrameworkElement` since the Xaml compiler expects to be able to use  the `FrameworkElement.Resources` property to get its resource dictionary.  Since `Window` isn't a `FrameworkElement`, the generated code breaks down.

Instead, you should move the UI logic (`<Grid RowDefinitions="Auto, *" x:Name="root">` etc.) into its own `UserControl`, and then set that `UserControl` as the content of the `Window`.

**Comment by michael-hawker** [comment]
Yeah, I too keep getting tripped up that Window isn't a `DependencyObject`... it's a bit weird.

If the expectation that `Window` is supposed to just be the host like the implicit `Frame` created in `App.xaml.cs` for UWP apps, maybe the File -> New template for WinUI 3 should separate out the `MainWindow` as also having a `MainPage` class hosted by it already setup?

This would help lead developers to the proper constructs when building their app and set them up for success vs. having everyone start dumping UI in `MainWindow` and then getting frustrated.

I mean I really like how the `MainWindow` is more exposed compared to all the extra logic on Startup that was done in the UWP space, but it's super confusing and bare-bones template when trying to start building an app from afterwards without creating a new `Page` first.

**Comment by Wufus** [comment]
> This would help lead developers to the proper constructs when building their app and set them up for success vs. having everyone start dumping UI in `MainWindow` and then getting frustrated.
> 
> I mean I really like how the `MainWindow` is more exposed compared to all the extra logic on Startup that was done in the UWP space, but it's super confusing and bare-bones template when trying to start building an app from afterwards without creating a new `Page` first.

But we have another problem: CustomTitleBar we can set only with XAML code in MainWindow.xaml, yes?
Using other UI in other page.xaml file will say No for united title with UI.

Best issue about i want to say #2124 with a lot of scrennsots

**Comment by qJake** [comment]
Coming from a strictly WPF/.NET Framework background, and trying out WinUI 3.x for the first time... this is really non-intuitive, especially when trying to port multi-window applications.

I realize that there is a fundamental difference in how UWP apps behave versus how traditional multi-window applications behave, but there should be an elegant solution to this problem nonetheless.

**Comment by KWodarczyk** [comment]
> Thanks for the report. The root of your `.xaml` needs to be a `FrameworkElement` since the Xaml compiler expects to be able to use the `FrameworkElement.Resources` property to get its resource dictionary. Since `Window` isn't a `FrameworkElement`, the generated code breaks down.
> 
> Instead, you should move the UI logic (`<Grid RowDefinitions="Auto, *" x:Name="root">` etc.) into its own `UserControl`, and then set that `UserControl` as the content of the `Window`.



> Thanks for the report. The root of your `.xaml` needs to be a `FrameworkElement` since the Xaml compiler expects to be able to use the `FrameworkElement.Resources` property to get its resource dictionary. Since `Window` isn't a `FrameworkElement`, the generated code breaks down.
> 
> Instead, you should move the UI logic (`<Grid RowDefinitions="Auto, *" x:Name="root">` etc.) into its own `UserControl`, and then set that `UserControl` as the content of the `Window`.

Isn't the Grid the root element ? the converter is defined in grid resources

**Comment by lavinders** [comment]
Can a sample / example be provided in WinUI docs so people don't spend cycles figuring out why this simple thing is not working. I am trying to use a boolean to visibility converter and it is so time consuming this simple thing.

**Comment by krschau** [comment]
> Can a sample / example be provided in WinUI docs

@gabbybilka

**Comment by thecosmicmuffet** [comment]
> > Thanks for the report. The root of your `.xaml` needs to be a `FrameworkElement` since the Xaml compiler expects to be able to use the `FrameworkElement.Resources` property to get its resource dictionary. Since `Window` isn't a `FrameworkElement`, the generated code breaks down.
> > Instead, you should move the UI logic (`<Grid RowDefinitions="Auto, *" x:Name="root">` etc.) into its own `UserControl`, and then set that `UserControl` as the content of the `Window`.
> 
> > Thanks for the report. The root of your `.xaml` needs to be a `FrameworkElement` since the Xaml compiler expects to be able to use the `FrameworkElement.Resources` property to get its resource dictionary. Since `Window` isn't a `FrameworkElement`, the generated code breaks down.
> > Instead, you should move the UI logic (`<Grid RowDefinitions="Auto, *" x:Name="root">` etc.) into its own `UserControl`, and then set that `UserControl` as the content of the `Window`.
> 
> Isn't the Grid the root element ? the converter is defined in grid resources

I think this is a worthwhile point. The auto-gen code successfully finds a Grid at the element where the resource is placed. It then passes '*this' (which is the Window) to SetConverterLookupRoot. When it could just set it as the target.as<Grid> that it's using for the SetDataRoot call. It's not a documentation issue. It's a autogen code issue with an inconsistent use of the term 'root' in the code block. e.g.:
 ```
 case 3: // MainWindow.xaml line 45
                {
                    auto element3 = target.as<::winrt::Microsoft::UI::Xaml::Controls::Grid>();
                    auto objBindings = std::make_unique<MainWindow_obj3_Bindings>();
                    objBindings->SetDataRoot(element3.DataContext()); // Here we are using the frameworkElement we found correctly as the 'dataRoot'
                    objBindings->SetConverterLookupRoot(*this); // Here we are inconsistently reverting to making the ConverterLookupRoot use *this, which is the Window
```

Refactoring code to avoid this autogen issue or making it about documentation seems like it's not very customer-focused. Why do we need to read more and refactor code, when this seems like a minor, and easily changeable portion of the generated code?

**Comment by GuildOfCalamity** [comment]
This is a big problem, how did you remove the ease of functionality that was binding in WPF/UWP and replace it with this?? 
I'm trying to use a boolean to visibility converter in my current project and cannot get it to work:
**Argument 1: cannot convert from 'Demo.MainWindow' to 'Microsoft.UI.Xaml.FrameworkElement'**
even though my converter and control is part of a <Grid> element.
I'm definitely not going to make some custom control as a work-around to this issue, it needs to be addressed ASAP!

**Comment by ACGNnsj** [comment]
This is definitely a problem. Why must I create a FrameworkElement instead of a Window?

**Comment by Konstantin-tr** [comment]
Since this is still not working, is there any plan to address this? Apart from this not being mentioned anywhere except for these bug reports, this is insanely impractical and the mentioned workaround is nothing short of a hack...

**Comment by Reza-Noei** [comment]
Hello Guys, Any plan to solve this issue ?

**Comment by GuildOfCalamity** [comment]
@Reza-Noei Unfortunately I don't think MS views this as a problem, and there will never be a "fix" since they have abstracted the `Window` object away from the `Microsoft.UI.Xaml.FrameworkElement`. I would suggest just using the **MainWindow** as a wrapper for a `Frame`/`Page` and then proceed as usual from there.

**Comment by Reza-Noei** [comment]
@GuildOfCalamity  ty for the answer. Sad for Microsoft 😢.
So they should remove it from default Template of the WinUI-3. That's so ridiculous 🤦‍♂️

**Comment by tinmac** [comment]
@RealTommyKlein 

The work around suggests using a UserControl but I need to use a bool to vis converter in the **TitleBar**

The **TitleBar** cant take a UserControl so how can we do this?

**Comment by neelabo** [comment]
The more familiar you are with WPF, the more confused you will be.
Even if the error is due to a different base architecture, this error message gives no information for a solution.
I finally managed to get to the information here from the vast web, but those who can't get to it are leaving WinUI3.
This is very problematic.

**Comment by CodyLyttle** [comment]
This is very unintuitive for anyone familiar with WPF.
One option is to use traditional binding when a converter is required.

```
<Grid x:Name="Root">
    <TextBlock Text="{Binding Text, Converter={StaticResource MyConverter}}"/>
</Grid>

// Code Behind
Root.DataContext = vm;
```

**Comment by rdeetz** [comment]
++ To all this. I am having this problem as well. It seems like both the generated code and the default template should be fixed.

**Comment by corned-beefhash** [comment]
5 years later this is still an issue

**Comment by Slion** [comment]
> One option is to use traditional binding when a converter is required.

Yeah but then you only get one time binding not one way or two way.

**Comment by michael-hawker** [comment]
@MartinAlejandroSoriano you should just encapsulate your UI in a Page or other control instead to continue using x:Bind over falling back to Binding.

**Comment by dotMorten** [comment]
@alexdi220 would function binding work for you? Generally with x:bind that'll work much better, avoid casting, give you type safety and better AoT compatibility. 
Example: Add this method to a class (in this case MainWindow but can be any class):

`public static string ToUpper(string value) => value.ToUpper();`

And then in Xaml on the window use function binding:

`Title="{x:Bind local:MainWindow.ToUpper(ViewModel.Title), Mode=OneWay}" >`

You could of course create a helper class with all your converter functions if you use it in more places than MainWindow, and then you don't need to implement all the IValueConverters.

**Comment by alexdi220** [comment]
> [Alexander Didenko (@alexdi220)](https://github.com/alexdi220) would function binding work for you? Generally with x:bind that'll work much better, avoid casting, give you type safety and better AoT compatibility. Example: Add this method to a class (in this case MainWindow but can be any class):
> 
> `public static string ToUpper(string value) => value.ToUpper();`
> 
> And then in Xaml on the window use function binding:
> 
> `Title="{x:Bind local:MainWindow.ToUpper(ViewModel.Title), Mode=OneWay}" >`
> 
> You could of course create a helper class with all your converter functions if you use it in more places than MainWindow, and then you don't need to implement all the IValueConverters.

Wow, five years have passed. I haven’t used WinUI since then, so I can’t say whether this solution would work or not.

In the original issue, the main point was the confusing and non-working API. I was passing a converter and getting an error. If converters were not supposed to be used with `x:Bind`, I would have expected the corresponding API/properties to be hidden from users as well.

So, I think it’s perfectly fine for `x:Bind` to take a different approach from the original WPF implementation, but in that case, this API should probably be removed.

**Comment by dotMorten** [comment]
@alexdi220 It's not so much that it isn't supported. If you create a page or user control, it'll work there. The real cause is an issue with the XAML code generation that only affects when the expression is directly in the Window. The above was a suggestion to get around the issue for now.

And yes, better late than never ;-)

<img width="1106" height="663" alt="Image" src="https://github.com/user-attachments/assets/269f9c16-a398-4a51-9c96-05673428d3de" />
